GSoC/GSoC-2020.md

```diff
@@ -0,0 +1,58 @@
+# Google Summer of Code 2020
+
+This is a list of project ideas for students applying to the Google Summer
+of Code program in 2020 (GSoC 2020).
+
+It's only an indicative list, and is subject to change as Godot is very
+actively developed by hundreds of contributors. Use this list as an
+inspiration source, but feel free to come with your own ideas on what you
+would like to work on in Godot.
+
+As a complement, you can check out the [community proposals repository](https://github.com/godotengine/godot-proposals)
+and fish out some good ideas for projects. You can also engage the discussions over there.
+
+Make sure to discuss your ideas with Godot developers on the #godotengine-devel
+IRC channel (Freenode server), so that we can help you refine your ideas
+and ensure that they fit in the Godot roadmap and architecture.
+
+## Reimplement Godot's theme editor
+
+**Skills:** C++, (potentially) GDScript
+
+Godot's theming system is very simple—it is still a game engine ;)—Styleboxes are assigned to different states of Control-Nodes. There is no cascading and no shared variables. This makes it very easy to understand, and results in very little complexity. But also comes with a cost. Setting up themes is super tedious. For example, a color change requires you to update hundreds of values.
+
+The goal of this project is to design a theme editor, which will generate a theme with systems, like a color palette, so that we get the best of both worlds. The simplicity of Godot's theming and an advanced theming structure when setting up a theme using the theme editor. There is a big amount of freedom and space for creative ideas to make creating themes a pleasure, while still keeping the core architecture as simple as possible.
+
+Another related task is to improve some aspects of the editor theme, such as caching icons for faster startup time, color icons with modulation (looks better on light theme, no re-render of icons when changing from dark to light), and more.
+
+## Multiplayer Scene Editor
+
+**Skills:** Networking, C++
+
+An editor plugin that allows multiple people to edit the same scene, using the remote scene edition protocol we have for running games. Possible features or issues:
+
+* A central server that provides authority for the edition protocol, maybe simlpy by forwarding messages in an arbitrary order.
+* Integration with VCS (such as Git, SVN, Mercurial, etc.)
+* Resolution of conflict when a new resource is added that requires new files (i.e. someone adds a texture that other people don't have. Maybe only allow new references to files that already exist in the VCS for simplicty, or provide warnings/placeholders when a client doesn't have the file).
+
+## Improvement in i18n for games
```

Comment:
I think it'd be interesting to mention support for [pluralization](https://github.com/godotengine/godot/issues/31432) here.

Also, is the performance difference between PO and MO files noticeable on a medium-sized project? If not, it may be difficult to justify the added complexity.

Reply:
Pluralization is mentioned already, I'll add a link to the issue.

> Also, is the performance difference between PO and MO files noticeable on a medium-sized project (say, 2,000 strings)? If not, it may be difficult to justify the added complexity. That doesn't mean it can't be added; I just consider it to be lower-priority compared to features like pluralization.

MO should be quite smaller, which is already a load optimization. Also, since it's binary, I assume it's actually easier to parse because it needs a rigid structure. I do agree that translation features are more important than formats, but they usually go hand-in-hand too.

Maybe we don't even want gettext. If there's a better more modern format which supports more i18n features we may actually want it instead.

Reply:
> Maybe we don't even want gettext. If there's a better more modern format which supports more i18n features we may actually want it instead.

I think the main established alternative out there is XLIFF. It tends to be more common in proprietary software circles and is more verbose overall – it's not fun to edit with a text editor. So, I think we can keep gettext for now.